### PR TITLE
[SPARK-14035][MLLIB] Make error message more verbose for mllib NaiveBayesSuite

### PR DESCRIPTION
**Pull request by jkbradley**
## What changes were proposed in this pull request?

Print more info about failed NaiveBayesSuite tests which have exhibited flakiness.
## How was this patch tested?

Ran locally with incorrect check to cause failure.

### CODE REVIEW

**Comment by jkbradley** [comment]
When I caused a failure by modifying the checks in the test, it gave an error message like this.  Note this is to verify the error message; it is not an actual test failure example:
```
[info]   Expected 0.5 and 1000000.4989503149 to be within 0.05 using absolute tolerance.
[info]   validateModelFit:
[info]    piData: [-0.6931471805599453, -1.2039728043259361, -1.6094379124341003]
[info]    thetaData: [-0.6931471805599453, -3.912023005428146, -3.912023005428146, -3.912023005428146, -3.912023005428146, -3.912023005428146, -3.912023005428146, -3.912023005428146, -3.912023005428146, -3.912023005428146, -3.912023005428146, -0.916290731874155]
[info]   [-3.912023005428146, -0.35667494393873245, -2.3025850929940455, -3.912023005428146, -3.912023005428146, -3.912023005428146, -3.912023005428146, -3.912023005428146, -3.912023005428146, -3.912023005428146, -3.912023005428146, -3.912023005428146]
[info]   [-3.912023005428146, -3.912023005428146, -0.5108256237659907, -3.912023005428146, -3.912023005428146, -3.912023005428146, -3.912023005428146, -3.912023005428146, -3.912023005428146, -3.912023005428146, -3.912023005428146, -1.2039728043259361]
[info]    model.labels: [0.0, 1.0, 2.0]
[info]    model.pi: [-0.6952487575155715, -1.2129771999650796, -1.5909161132025096]
[info]    model.theta: [-0.6967594662458589, -3.7882040913369224, -3.9722971277792594, -3.753417975251507, -3.8334606829250433, -3.8521528159371963, -3.86163155989174, -3.930624431378691, -3.8808629218196273, -4.0269555403171235, -3.930624431378691, -0.9227216222044454]
[info]   [-3.778491612803623, -0.35931949410361774, -2.174953422496711, -3.823612048084093, -3.823612048084093, -3.5912800707154764, -3.9549480501451795, -3.8871254538064184, -3.778491612803623, -4.126798307071839, -4.106179019869103, -4.027707404427607]
[info]   [-3.8360251366561937, -3.63123072401018, -0.4893159402781082, -4.036695832118345, -3.7700671688643963, -3.906642703870147, -4.219017388912299, -3.8825451522910863, -3.9313353164605185, -3.7700671688643963, -3.813552280804135, -1.2166405726396397] (TestingUtils.scala:79)
```

**Comment by jkbradley** [comment]
CC: @mengxr Seem reasonable?

File: mllib/src/test/scala/org/apache/spark/mllib/classification/NaiveBayesSuite.scala [hunk 1]
```diff
@@ -103,17 +104,26 @@ class NaiveBayesSuite extends SparkFunSuite with MLlibTestSparkContext {
...
-    for (i <- modelIndex) {
-      for (j <- 0 until thetaData(i._2).length) {
-        assert(closeFit(math.exp(thetaData(i._2)(j)), math.exp(model.theta(i._1)(j)), 0.05))
+    val modelIndex = piData.indices.zip(model.labels.map(_.toInt))
+    try {
+      for (i <- modelIndex) {
+        assert(math.exp(piData(i._2)) ~== math.exp(model.pi(i._1)) absTol 0.05)
+      }
+      for (i <- modelIndex) {
+        for (j <- thetaData(i._2).indices) {
```
Comment (srowen) [inline mllib/src/test/scala/org/apache/spark/mllib/classification/NaiveBayesSuite.scala]: +1 for `.indices`; while nit-picking these two for statements could be merged into one.

File: mllib/src/test/scala/org/apache/spark/mllib/classification/NaiveBayesSuite.scala [hunk 2]
```diff
@@ -103,17 +104,26 @@ class NaiveBayesSuite extends SparkFunSuite with MLlibTestSparkContext {
...
+          assert(math.exp(thetaData(i._2)(j)) ~== math.exp(model.theta(i._1)(j)) absTol 0.05)
+        }
       }
+    } catch {
+      case e: TestFailedException =>
+        def arr2str(a: Array[Double]): String = a.mkString("[", ", ", "]")
+        val msg = "\nvalidateModelFit:\n" +
```
Comment (srowen) [inline mllib/src/test/scala/org/apache/spark/mllib/classification/NaiveBayesSuite.scala]: You could also make a function that generates the error message you want, and pass that as an arg to the asserts. It's more straightforward but not sure it's simpler, so whatever you like here.

**Comment by srowen** [comment]
Merged to master (though leaving JIRA open of course)

**Comment by jkbradley** [comment]
Thanks!

**Pull request closed without merging**